### PR TITLE
PLENTY-427 category filter on category pages fix

### PR DESCRIPTION
**Pull request by msilinskas**

### CODE REVIEW

File: resources/js/src/app/components/itemList/filter/ItemCategoryDropdown.js [hunk 1]
```diff
@@ -19,12 +24,24 @@ Vue.component("item-category-dropdown", {
...
+            return typeof this.getSelectedFilters().find(element => element.id == this.facet.id) !== 'undefined';
+        },
     },
 
     methods: {
         getSubCategoryName(parentCategory, subCategory) {
             return parentCategory.name + '_' + subCategory.name;
+        },
+
+        categorySelected(category) {
```
Comment (TheKeymaster) [inline resources/js/src/app/components/itemList/filter/ItemCategoryDropdown.js:43]: Why would you declare this method? It is the same as `isCategorySelected`.

File: resources/js/src/app/mixins/url.js [hunk 2]
```diff
@@ -277,6 +277,27 @@ export default {
             return selectedFilters;
         },
 
+        /**
+         * @param category
```
Comment (TheKeymaster) [inline resources/js/src/app/mixins/url.js:281]: Please add a typehint.

File: resources/js/src/app/mixins/url.js [hunk 3]
```diff
@@ -277,6 +277,27 @@ export default {
...
+         * @returns {boolean}
+         */
+        isCategorySelected(category) {
+            let selectedFilters = this.getSelectedFilters(),
+                splitedSelectedCategories = [];
```
Comment (TheKeymaster) [inline resources/js/src/app/mixins/url.js:286]: Define variables separately on a separate line from eachother.

File: resources/js/src/app/mixins/url.js [hunk 4]
```diff
@@ -277,6 +277,27 @@ export default {
             return selectedFilters;
         },
 
+        /**
+         * @param category
+         * @returns {boolean}
+         */
+        isCategorySelected(category) {
+            let selectedFilters = this.getSelectedFilters(),
```
Comment (TheKeymaster) [inline resources/js/src/app/mixins/url.js:285]: This could be a `const`.

File: resources/js/src/app/mixins/url.js [hunk 5]
```diff
@@ -277,6 +277,27 @@ export default {
...
+                splitedSelectedCategories = [];
+
+            for (let i = 0; i < selectedFilters.length; i++) {
+                if (selectedFilters[i].id !== this.facet.id) {
+                    continue;
+                }
+
+                splitedSelectedCategories = selectedFilters[i].name.split('>');
```
Comment (TheKeymaster) [inline resources/js/src/app/mixins/url.js:293]: Wouldn't this cause a conflict when multiple categories have the same name? E.g.

* Men > Shoes
* Women > Shoes

File: tests/Api/Request/ParametersBuilderTest.php [hunk 6]
```diff
@@ -83,7 +83,35 @@ public function setSearchParamsProvider()
                         0 => 'variation_id'
                     ],
                     'attrib' => [
-                        'color' => ['red', 'blue']
+                        'color' => ['red', 'blue'],
+                        'cat' => ['Category 2']
```
Comment (TheKeymaster) [inline tests/Api/Request/ParametersBuilderTest.php:87]: Why was this added? The other parameters make sense, but why do we need a category now?

File: tests/Api/Request/ParametersBuilderTest.php [hunk 7]
```diff
@@ -83,7 +83,35 @@ public function setSearchParamsProvider()
...
+                    ],
+                    'selected' => ['cat' => ['Category']],
+                    'order' => 'price ASC',
+                    'count' => 0,
+                    'first' => 0
+                ]
+            ],
+            'Category page request with subcategory selected' => [
+                'parameters' => [
+                    Plugin::API_PARAMETER_ATTRIBUTES => [
```
Comment (TheKeymaster) [inline tests/Api/Request/ParametersBuilderTest.php:97]: Isn't `attrib` as a string used everywhere else? I typically prefer to not use constants in unit-tests, as if the constant changes in the future, you can see which components would be affected.

File: resources/js/src/app/components/itemList/filter/ItemFilter.js [hunk 8]
```diff
@@ -63,5 +67,9 @@ Vue.component("findologic-item-filter", {
                 name: parentCategory.name + '_' + subCategory.name
             };
         },
+
+        categorySelected(category) {
```
Comment (TheKeymaster) [inline resources/js/src/app/components/itemList/filter/ItemFilter.js:71]: Please add a return typehint.

File: resources/js/src/app/components/itemList/filter/ItemCategoryDropdown.js (hunk 1 above)
Comment (TheKeymaster) [inline resources/js/src/app/components/itemList/filter/ItemCategoryDropdown.js:43]: Please add a return typehint.

File: resources/js/src/app/components/itemList/filter/ItemCategoryDropdown.js [hunk 9]
```diff
@@ -19,12 +24,24 @@ Vue.component("item-category-dropdown", {
             }
 
             return label;
-        }
+        },
+
+        isSelected() {
```
Comment (TheKeymaster) [inline resources/js/src/app/components/itemList/filter/ItemCategoryDropdown.js:29]: Please add a return typehint.

File: resources/js/src/app/components/itemList/filter/ItemCategoryDropdown.js [hunk 10]
```diff
@@ -19,12 +24,24 @@ Vue.component("item-category-dropdown", {
...
+            if (this.facet.id === 'cat' && this.facet.values.length === 1) {
+                return true;
+            }
+
+            return typeof this.getSelectedFilters().find(element => element.id == this.facet.id) !== 'undefined';
```
Comment (TheKeymaster) [inline resources/js/src/app/components/itemList/filter/ItemCategoryDropdown.js:34]: Why aren't you using strict comparison when checking the id?
```suggestion
            return typeof this.getSelectedFilters().find(element => element.id === this.facet.id) !== 'undefined';
```

**Review by TheKeymaster — CHANGES_REQUESTED** [review]

File: resources/js/src/app/components/itemList/filter/ItemCategoryDropdown.js [hunk 11]
```diff
@@ -7,6 +7,11 @@ Vue.component("item-category-dropdown", {
     computed: {
         dropdownLabel() {
             let selectedFilters = this.getSelectedFilters();
+
+            if (selectedFilters.length === 0 && this.facet.id === 'cat' && this.facet.values.length === 1) {
```
[the remark this answers is not in the document]
Comment (msilinskas) [reply resources/js/src/app/components/itemList/filter/ItemCategoryDropdown.js:11]: We need this check for the default label setting in the category page.

File: tests/Api/Request/ParametersBuilderTest.php (hunk 6 above)
Comment (msilinskas) [reply tests/Api/Request/ParametersBuilderTest.php:87]: We have commented out the continue in the [src/Api/Request/ParametersBuilder](https://github.com/findologic/plugin-plentymarkets-ceres-api/pull/303/files#diff-e93f58cf887da50267647fd05d8598121210af7bb2924259e16341590c69d567)::87. Now we are setting cat attribute to the request, so I had to add it here too.

File: resources/js/src/app/components/itemList/filter/ItemFilter.js [hunk 12]
```diff
@@ -63,5 +69,19 @@ Vue.component("findologic-item-filter", {
...
+        /**
+         * @returns {boolean}
+         */
+        shouldShowFilter() {
+            return this.facet.id !== 'cat' ||
+                this.facet.id === 'cat' && typeof this.showCategoryFilter === 'undefined' ||
+                this.facet.id === 'cat' && this.showCategoryFilter;
+        },
+
+        shouldShowCategoryFilter() {
```
Comment (TheKeymaster) [inline resources/js/src/app/components/itemList/filter/ItemFilter.js:76]: Great that you also added typehints to most methods. Here you could also add one :)

File: src/Containers/SearchFilterContainer.php [hunk 13]
```diff
@@ -19,12 +21,20 @@ public function call(Twig $twig, SearchService $searchService): string
         }
 
         $searchResults = $searchService->getResults();
+        $currentCategory = $searchService->getCategoryService()->getCurrentCategory();
+        $showCategoryFilter = true;
+
+        if (null !== $currentCategory && $currentCategory->level !== 1) {
```
Comment (TheKeymaster) [inline src/Containers/SearchFilterContainer.php]: Please don't use [yoda conditions](https://en.wikipedia.org/wiki/Yoda_conditions).

```suggestion
        if ($currentCategory !== null && $currentCategory->level !== 1) {
```

Comment (TheKeymaster) [reply src/Containers/SearchFilterContainer.php]: Maybe you wanna change the condition to `> 1` to make it more clear that everything above should not be shown. Also a comment might be helpful, otherwise someone reading this may be confused, why the category filter is not shown further levels below.

File: resources/js/src/app/mixins/url.js (hunk 5 above)
Comment (TheKeymaster) [reply resources/js/src/app/mixins/url.js:293]: Is this still an issue? If not, please resolve the conversation.

File: resources/js/src/app/components/itemList/filter/ItemFilter.js [hunk 14]
```diff
@@ -63,5 +69,19 @@ Vue.component("findologic-item-filter", {
                 name: parentCategory.name + '_' + subCategory.name
             };
         },
+
+        /**
+         * @returns {boolean}
+         */
+        shouldShowFilter() {
```
Comment (TheKeymaster) [inline resources/js/src/app/components/itemList/filter/ItemFilter.js]: Can't `shouldShowFilter` and `shouldShowCategoryFilter` be combined somehow?

**Review by TheKeymaster — CHANGES_REQUESTED** [review]

File: resources/js/src/app/components/itemList/filter/ItemFilter.js (hunk 14 above)
Comment (msilinskas) [reply resources/js/src/app/components/itemList/filter/ItemFilter.js]: I've moved the checking if facet id is equals to cat to the twig file and left only shouldShowCategoryFilter method in the js component.

**Review by TheKeymaster — APPROVED** [review]
